File: internal/learning/article.go
package learning

import (
	"context"
	"fmt"
	"xuetangx/internal/domain"
	"xuetangx/internal/platform/wire"
)

func (r *Runner) article(ctx context.Context, in Input, u domain.Unit) (string, error) {
	leaf, e := r.leaf(ctx, in.Course, u)
	if e != nil {
		return "", e
	}
	if wire.True(leaf["finish"]) {
		return "skipped", nil
	}
	sku, e := wire.ID(leaf["sku_id"])
	if e != nil {
		return "", e
	}
	if _, e = r.request(ctx, "GET", fmt.Sprintf("/api/v1/lms/learn/user_article_finish/%d/?cid=%d&sid=%d", u.ID, in.Course.ClassroomID, sku), nil); e != nil {
		return "", e
	}
	e = r.verify(ctx, func() (bool, error) { d, e := r.leaf(ctx, in.Course, u); return wire.True(d["finish"]), e })
	return "completed", e
}

File: server/workflow/workers/article.js
const { completeValue, failure, requireId } = require('./common');
function createAction({ rpc, request, leaf, verify }) {
  async function article(course, unit) {
    const data = await leaf(course, unit); if (completeValue(data.finish)) return { status: 'skipped' };
    await request('GET', `/api/v1/lms/learn/user_article_finish/${unit.id}/?${new URLSearchParams({ cid: course.classroomId, sid: requireId(data.sku_id) })}`);
    if (!await verify(async () => completeValue((await leaf(course, unit)).finish))) throw failure('后端尚未确认图文完成', 'UNCONFIRMED');
    return { status: 'completed' };
  }
  return article;
}
module.exports = { createAction };
